resourcet/Data/Acquire/Internal.hs

```diff
@@ -25,8 +27,21 @@ import qualified Control.Monad.Catch as C ()
 -- @since 1.1.2
 data ReleaseType = ReleaseEarly
                  | ReleaseNormal
-                 | ReleaseException
-    deriving (Show, Read, Eq, Ord, Enum, Bounded, Typeable)
+                 | ReleaseException' E.SomeException
+    deriving (Show, Typeable)
+
+-- | Fake 'E.Exception' to use with the deprecated 'ReleaseException' pattern.
+data DeprecatedReleaseExceptionPlaceholder = DeprecatedReleaseExceptionPlaceholder
+     deriving (Show)
+
+instance E.Exception DeprecatedReleaseExceptionPlaceholder
+
+{-# COMPLETE ReleaseEarly, ReleaseNormal, ReleaseException #-}
+{-# DEPRECATED ReleaseException "Use ReleaseException'" #-}
+pattern ReleaseException :: ReleaseType
+pattern ReleaseException <- ReleaseException' _
+  where
+    ReleaseException = ReleaseException' (E.toException DeprecatedReleaseExceptionPlaceholder)
```

Comment:
This is a bidirectional pattern synonym, meaning it can be used both for *constructing* a value and *pattern matching* on it.

The first line:

```haskell
pattern ReleaseException <- ReleaseException' _
```

describes how the pattern match works. When you match on `ReleaseException`, it will 'replace' that with a match on `ReleaseException' _` - this catch-all just ignores the new parameter.

Reply:
The second part,

```haskell
  where
    ReleaseException = ReleaseException' (E.toException DeprecatedReleaseExceptionPlaceholder)
```

describes what happens when you use this to *construct* with the term. Let's consider this definition:

```haskell
foo :: ReleaseType
foo = ReleaseException
```

Before this code change, we'd have a simple value - `ReleaseException :: ReleaseType`. After this change, the syntax doesn't change, but the actual value that gets constructed becomes `ReleaseException' (SomeException DeprecatedReleaseExceptionPlaceholder)`.

---

I'm unsure how I feel about having the pattern synonym for construction as well as matching. Constructing a `ReleaseType` should really rarely happen, and as a library user, I'd *much* rather have an error pointing me towards `ReleaseExceptionWith`, so I can update my code and provide the exception.

Right now, that notice is given using the `DEPRECATION` pragma on `ReleaseException`, pointing us towards the new `ReleaseException'`. But this means that you'll get warnings (and `-Werror` build failures) when pattern matching on `ReleaseException`, which is most likely much more common for end-users to do than constructing.

If we drop the construction half of the synonym, then we don't need to deprecate the pattern synonym, and I'd guess almost all end users will have no new warnings or errors. The downside of this is that end users won't know about this change unless they go looking for it.

Briefly, the two choices appear to be:

- All users that reference the constructors get `DEPRECATION` warnings
- Users that *construct* a `ReleaseException` get a compile-time error, and users that match on a `ReleaseException` continue on as-usual.

Reply:
There's also:

* Users that *construct* a `ReleaseException` get a compile-time error, and users that match on a `ReleaseException` get a `DEPRECATION` warning

I see the argument for making the pattern uni-directional, but it's not clear why we still wouldn't want users to move to the proper constructor in matching.

Reply:
I guess there are two more options:

- Users don't get any errors or warnings (no `DEPRECATION` message + bidirectional synonym)
- Users get compile-time errors for pattern matching or construction (no `PatternSynonym` at all)

From a "no breaking changes" perspective, "no errors or warnings" satisfies it pretty well - lacking the deprecation warning, this variant isn't even a major breaking change.

But it has a cost - the end-user will not be aware of two things:

1. If the end user is constructing `ReleaseException`s, then they will continue to (unknowingly) construct dummy values that drop useful information.
2. If the end user is *matching* on `ReleaseException`, and they want to know about the exception, then they won't know about it.

I'm more worried about the downstream effects of 1. Having to worry about dummy exceptions is really annoying, all so some weird intermediate library scenario can continue to sort-of have backwards compatibility.

On the other hand, "no pattern synonym at all" means that everyone gets a compile error. This means that folks will be made aware of these two bits of information as they fix their build. BUT, they have to fix their code. And then you have the whole major version hullabaloo with CPP to support multiple versions.

---

So, here's a middle way:

1. Introduce the bidirectional pattern synonym, *without a deprecation*, as a *minor* version bump. Since we're not introducing new warnings or errors, and pattern matches will all be `COMPLETE` still, I think this is fine. This version bump can be a 'compatibility' version.
2. Immediately afterwards, deprecate the `ReleaseException` pattern synonym with an upgrade path message. This is a major version bump.
3. At the *next* major version, delete the synonym and other deprecated terms.

With this, end users can depend on `resourcet-1.2.7` (next minor version). Then `resourcet-1.3.0` has the warnings, and `resourcet-1.4.0` removes the synonym entirely.

Libraries that require `resourcet <= 1.3` won't block end users from getting access to the exception, so we can deliver the feature without blocking anyone.

Reply:
All of these "bad case" scenarios are pretty unlikely though. I think any solution with a pattern synonym here will prevent any major pain or churn in the ecosystem.

Reply:
> lacking the deprecation warning, this variant isn't even a major breaking change.

I don't think a deprecation should move a minor bump to a major one.

Reply:
> And then you have the whole major version hullabaloo with CPP to support multiple versions.

Nah, just wildcard match if you want to support both types and don't care about the exception.